### ISSUE TITLE
[BUG] Extra range not working with DataRange1d

### ISSUE THREAD

**Issue by muendlein** [bokeh/bokeh#13297]
### Software versions

IPython version     :  (not installed)
Tornado version     :  6.2
Bokeh version       :  3.2.1
node.js version     :  v16.15.0
npm version         :  8.5.5
Operating system    :  Windows-10-10.0.19045-SP0

### Expected behavior

Using extra x/y ranges should work with both DataRange1d and Range1d even if the extra data is added dynamically.

### Observed behavior

When adding the extra data dynamically (i.e. with a button click (see example)), a JS error (`[bokeh] could not set initial ranges`) will be thrown in case of DataRange1d. This does not happen for Range1d with fixed ranges.

### Example code

```Python
from bokeh.io import curdoc
from numpy import arange, linspace, pi, sin

from bokeh.layouts import column
from bokeh.models import (CustomJS, LinearAxis, Range1d, Select,
                          WheelZoomTool, ZoomInTool, ZoomOutTool, DataRange1d, Button)
from bokeh.palettes import Sunset6
from bokeh.plotting import figure, show

x = arange(-2*pi, 2*pi, 0.2)
y = sin(x)
y2 = linspace(0, 100, len(x))

blue, red = Sunset6[2], Sunset6[5]

p = figure(x_range=(-2*pi, 2*pi), y_range=(-1, 1), tools="pan,box_zoom,save,reset")
p.background_fill_color = "#fafafa"
p.axis.axis_label = "light blue circles"
p.axis.axis_label_text_color = blue
p.scatter(x, y, line_color="black", fill_color=blue, size=12)


def add_extra_data():
    p.extra_y_ranges['foo'] = DataRange1d() # Range1d(0, 100) works as intended

    ax2 = LinearAxis(
        axis_label="red circles",
        y_range_name="foo",
    )
    ax2.axis_label_text_color = red
    p.add_layout(ax2, 'right')

    p.scatter(x, y2, color=red, size=8,
                            y_range_name="foo",
                            )


button_add_data = Button(label="Click to add extra data")
button_add_data.on_click(add_extra_data)

doc = curdoc()
doc.add_root(column(button_add_data, p))
```

**Comment by bryevdv** [comment]
@muendlein I would write your code like this (which works as you describe)
```python
ax2 = LinearAxis(
    axis_label="red circles",
    y_range_name="foo",
)
ax2.axis_label_text_color = red
p.add_layout(ax2, 'right')

r = p.scatter(x, y2, color=red, size=8, y_range_name="foo")

ax2.visible = r.visible = False

def add_extra_data():
    ax2.visible = r.visible = True
```

This kind of pattern&mdash;adding things up front and then toggling their visibility, etc&mdash;is always considered the best practice for kind of situation. In general actually swapping in and out objects like this for real is a very heavyweight operation, both potentially in terms of actual data over the wire, but also definitely in terms of  document bookkeeping. 

cc @bokeh/dev

**Comment by muendlein** [comment]
> @muendlein I would write your code like this (which works as you describe)
> 
> ```python
> ax2 = LinearAxis(
>     axis_label="red circles",
>     y_range_name="foo",
> )
> ax2.axis_label_text_color = red
> p.add_layout(ax2, 'right')
> 
> r = p.scatter(x, y2, color=red, size=8, y_range_name="foo")
> 
> ax2.visible = r.visible = False
> 
> def add_extra_data():
>     ax2.visible = r.visible = True
> ```
> 
> This kind of pattern—adding things up front and then toggling their visibility, etc—is always considered the best practice for kind of situation. In general actually swapping in and out objects like this for real is a very heavyweight operation, both potentially in terms of actual data over the wire, but also definitely in terms of document bookkeeping.
> 
> cc @bokeh/dev

I fully agree that your pattern is preferred, but shouldn't my code still yield same the result unless there is a bug? If you do not consider this to be a bug, it would probably make sense to document that extra ranges must be set upfront.
